# Week – 3: Classifcation Model

## Problem Statement
Implement a machine learning classification algorithm on image to recognize
handwritten digits from a paper.

## Dataset Description

In this experiment, we use the **Digits dataset** available in `scikit-learn`.

It can be loaded using:

```python
from sklearn import datasets
digits = datasets.load_digits()
```

### Overview

* Total samples: **1797**
* Classes: **10 (digits 0–9)**
* Image size: **8 × 8 pixels**
* Problem type: **Multiclass classification**

### Feature Matrix (X)

Each image is converted into numerical form and avaliable when we are calling the .data

```python
X = digits.data
print(X.shape)
```

Output:

```
(1797, 64)
```

This means:

* There are **1797 images**
* Each image is represented using **64 features**

Why 64 features or attrbiutes or columns?

```
8 × 8 = 64
```

Each 8×8 image is flattened into a **64-dimensional feature vector (columns)**.

If we print one sample:

```python
print(digits.data[0])
```

Example Output:

```
[ 0.  0.  8. 16.  5.  0.  0.  0.  0.  1. 13. 11. 16.  0.  0.  0.
  0.  0. 10.  0. 13.  3.  0.  0.  0.  0.  3.  1. 16.  1.  0.  0.
  0.  0.  0.  9. 12.  0.  0.  0.  0.  0.  3. 15.  5.  0.  0.  0.
  0.  0. 14. 15.  8.  8.  3.  0.  0.  0.  7. 12. 12. 12. 13.  1.]
```

These values represent **pixel intensity values** (grayscale levels) and range values 0 and 16.

### Target (y)

```python
y = digits.target
print(y[0])
```

Output:

```
2
```

This means:

* Each value represents the actual digit label (0–9)
* `X[i]` corresponds to digit `y[i]`


### Viewing the Original Image

The dataset also stores images in 8×8 format.

To visualize the digit:

```python
import matplotlib.pyplot as plt
plt.imshow(digits.images[0], cmap='gray')
plt.show()
```

* `imshow()` means **image show**
* It converts a NumPy array into an image
* `cmap='gray'` is used because the dataset contains **grayscale images**

This dataset is used in this experiment to train an **SVM classifier** for handwritten digit classification.



## Procedure

In [ ]:
import numpy as np
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report

# Load MNIST dataset
digits = datasets.load_digits()

# Features and labels
X = digits.data
y = digits.target


# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Initialize SVM classifier
model = SVC(kernel='linear')

# Train model
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)

# Evaluate model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)
print(classification_report(y_test, y_pred))


Accuracy: 0.9777777777777777
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        33
           1       0.97      1.00      0.98        28
           2       1.00      1.00      1.00        33
           3       0.97      0.94      0.96        34
           4       0.98      0.98      0.98        46
           5       0.96      1.00      0.98        47
           6       1.00      1.00      1.00        35
           7       0.97      0.97      0.97        34
           8       1.00      0.97      0.98        30
           9       0.95      0.93      0.94        40

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



## Results

The SVM classifier with a **linear kernel** achieved an accuracy of **97.78%** on the test dataset.

From the classification report:

* Most digits have **precision, recall, and F1-score close to 1.00**.
* Digits such as **0, 2, and 6** were classified perfectly.
* Slight misclassifications occurred in digits like **3 and 9**, but performance is still very high.
* The **macro average and weighted average F1-score are 0.98**, indicating balanced performance across all classes.

This shows that:

* The dataset is well-structured and separable.
* A linear SVM is sufficient to classify most handwritten digits correctly.
* The model generalizes well on unseen test data.

Overall, the experiment demonstrates that SVM is highly effective for handwritten digit classification on the Digits dataset.
